### ISSUE TITLE
bail observed on ~zod on release candidate testnet

### ISSUE THREAD

**Issue by jtobin** [urbit/urbit#1322]
(Occurring on the release candidate HEAD, i.e. c1e2a479c4242c287372c0f67112e14b05644031.)

Checking on an instance of `~zod` that had been running on a testnet showed that it had bailed with the following:

```
~zod:dojo> allocate: reclaim: half of 82 entries
Assertion 'u3R->cap_p > u3R->hat_p' failed in noun/retrieve.c:1600

bail: oops
bailing out
urbit: noun/manage.c:663: u3m_bail: Assertion `0' failed.
                                                         Aborted (core dumped)
```

The scrollback contained a lot of miscellaneous `%ames`-related errors and stack traces, as well as repeated `bail: exits`.  You can see a representative sample [here](https://gist.github.com/jtobin/a7213db754825d277f755e4f1dc5f710) -- this sort of thing was repeated many times.

Relevant backtrace from the core:

```
Program terminated with signal SIGABRT, Aborted.
#0  0x00007ff6b0a47be0 in raise () from /nix/store/mrfcv8ipiksfdrx3xq7dvcrzgg2jdfsw-glibc-2.27/lib/libc.so.6
[Current thread is 1 (LWP 24128)]
(gdb) bt
#0  0x00007ff6b0a47be0 in raise () from /nix/store/mrfcv8ipiksfdrx3xq7dvcrzgg2jdfsw-glibc-2.27/lib/libc.so.6
#1  0x00007ff6b0a48e90 in abort () from /nix/store/mrfcv8ipiksfdrx3xq7dvcrzgg2jdfsw-glibc-2.27/lib/libc.so.6
#2  0x00007ff6b0a406e7 in __assert_fail_base () from /nix/store/mrfcv8ipiksfdrx3xq7dvcrzgg2jdfsw-glibc-2.27/lib/libc.so.6
#3  0x00007ff6b0a40792 in __assert_fail () from /nix/store/mrfcv8ipiksfdrx3xq7dvcrzgg2jdfsw-glibc-2.27/lib/libc.so.6
#4  0x0000000000451a23 in u3m_bail ()
#5  0x00000000004518ce in c3_cooked ()
#6  0x000000000044b4a0 in u3r_mug ()
#7  0x000000000043ee2c in u3h_put ()
#8  0x000000000041f5f0 in u3qe_cue ()
#9  0x000000000041faed in u3ke_cue ()
#10 0x000000000046e0a1 in _pier_work_poke ()
#11 0x000000000045546d in _newt_read_cb ()
#12 0x00000000004c34af in uv__read (stream=stream@entry=0x231ffc0) at src/unix/stream.c:1192
#13 0x00000000004c3b9c in uv__stream_io (loop=<optimized out>, w=0x2320048, events=1) at src/unix/stream.c:1259
#14 0x00000000004c8640 in uv__io_poll (loop=loop@entry=0x8d0140 <default_loop_struct>, timeout=13604) at src/unix/linux-core.c:380
#15 0x00000000004baa68 in uv_run (loop=0x8d0140 <default_loop_struct>, mode=UV_RUN_DEFAULT) at src/unix/core.c:354
#16 0x00000000004664ba in u3_daemon_commence ()
#17 0x0000000000409b51 in main ()
```

**Comment by joemfb** [comment]
The linked %ames-crash is caused by other ships sending packets to you with an incorrect life. These are likely out-of-date (or incorrectly-booted) testnet ships.

This crash is technically a `bail: meme`, ie OOM observed in `u3r_mug`'s explicit usage of the road stack.

**Comment by belisarius222** [comment]
What I don't get is why the ship is running out of memory while attempting to handle a %hole Ames crash event.  I would have thought that would basically just print a stack trace and make some minor adjustments to Ames state.

I see three possibilities here:

1) Something is deeply amiss in the error handling flow that causes way too much memory usage.  If I had to guess, I'd say it's somewhere in the printing logic.

2) The ship was almost out of memory to begin with, and this event was the straw that broke the camel's back.

3) The runtime does something uncouth on a %hole event that balloons the loom.

—
~rovnys-ricfer

https://urbit.org

**Comment by joemfb** [comment]
It's unclear from this report whether the event in question was the %ames %hole bad-packet flow, or something else.

More importantly, and I failed to notice this until now, this crash is a `bail: meme` in the daemon, *not* in the worker. In other words, it's in the process that has no persistent memory, only ephemeral memory related to processing events. So this is either a space leak or memory leak in the normal flow of event processing in the i/o runtime, not a leak inside Arvo.

**Comment by cgyarvin** [comment]
This would be a good excuse to replay the log, if it exists, with additional instrumentation.  I guess there is no obvious way to pull this trick in cc-release, but maybe there should be...

**Comment by belisarius222** [comment]
I wonder if there's a space leak in the daemon that caused it to hold on to old events in the loom. Can we run the equivalent of |mass on the daemon? It does have a loom ...

—
~rovnys-ricfer
https://urbit.org

**Comment by joemfb** [comment]
Replaying would be good. The daemon does less during replay, so it's more likely to succeed without hitting this failure condition. Instrumentation would be manual.

You can't run `|mass` in the daemon because of .... uh ... leaks.

**Comment by belisarius222** [comment]
A clue, Sherlock!

—
~rovnys-ricfer
https://urbit.org

**Comment by ixv** [comment]
We got another crash on the testnet, this time on ~nec, but we had |automass running this time.

**Comment by joemfb** [comment]
A core dump caused by a bail meme/oops in the daemon?

**Comment by joemfb** [comment]
If so, can you post the most reset `|mass` output for the daemon?

**Comment by ixv** [comment]
I'm not 100% sure how to check if the crash was in the daemon, but I do see the function call "u3_daemon_grab" and "u3_daemon_commence" in the stack trace:

```
urb@nec:~$ lldb -c core-old
(lldb) target create --core "core-old"
Core file '/home/urb/core-old' (x86_64) was loaded.
(lldb) bt
* thread #1, name = 'urbit-testnet', stop reason = signal SIGABRT
  * frame #0: 0x00007f3011825be0 libc.so.6`__GI_raise + 256
    frame #1: 0x00007f3011826e90 libc.so.6`__GI_abort + 528
    frame #2: 0x00007f301181e6e7 libc.so.6`__assert_fail_base + 279
    frame #3: 0x00007f301181e792 libc.so.6`__GI___assert_fail + 66
    frame #4: 0x0000000000451a63 urbit-testnet`u3m_bail + 339
    frame #5: 0x000000000045190e urbit-testnet`c3_cooked + 14
    frame #6: 0x000000000044444a urbit-testnet`u3a_sweep + 666
    frame #7: 0x000000000046661c urbit-testnet`u3_daemon_grab + 396
    frame #8: 0x00000000004c65a8 urbit-testnet`uv__run_timers(loop=0x00000000008d0140) at timer.c:165
    frame #9: 0x00000000004bac95 urbit-testnet`uv_run(loop=0x00000000008d0140, mode=UV_RUN_DEFAULT) at core.c:345
    frame #10: 0x000000000046641a urbit-testnet`u3_daemon_commence + 266
    frame #11: 0x0000000000409b81 urbit-testnet`main + 1313
    frame #12: 0x00007f3011812b8e libc.so.6`__libc_start_main + 238
    frame #13: 0x000000000040a62a urbit-testnet`_start at start.S:120
(lldb) ^D
```

I just looked at the latest |mass output for the daemon and its an empty file, but here is the output of the next most recent:
```
~2019.7.11..20.10.31..6357
  wish cache: MB/10.248.268
  date: B/36
  formatted date: B/48
  instance string: B/32
  kernel: KB/62.296
total arvo stuff: MB/10.310.680
  warm jet state: KB/32.144
  cold jet state: KB/21.072
  hank cache: B/704
  battery hash cache: B/288
  call site cache: B/660
  hot jet state: KB/127.424
total jet stuff: KB/182.292
  bytecode programs: KB/836.080
  bytecode cache: KB/52.808
total nock stuff: KB/888.888
  trace buffer: B/36
  memoization cache: KB/3.720
total road stuff: KB/3.756
  marked 4 writs
  events: KB/4.252
  encoded events: B/508
total pier stuff: KB/4.760
total marked: MB/11.390.376
sweep: MB/11.390.376
```

**Comment by joemfb** [comment]
This is the daemon, and it crashed due to a memory leak encountered when running `|mass`. Do you have the console output from the crash?

**Comment by ixv** [comment]
Yes and its an enormous mess of line noise, is there something in particular you're looking for?

**Comment by joemfb** [comment]
All of it. Gist or email is fine.

**Comment by joemfb** [comment]
BTW, the pier should be fine. You can restart and just keep going as is. If it crashes frequently enough to interfere with testing you can turn of |automass (assuming that's possible ;)).

**Comment by ixv** [comment]
https://gist.github.com/ixv/af8f62e5aad0ba38db4ecdbf1f5bda8c

**Comment by ixv** [comment]
I just tried rebooting that pier and it immediately bailed again:
```
urb@nec:~$ urbit nec
~
urbit 0.7.3
boot: home is /home/urb/nec
loom: mapped 2048MB
boot: protected loom
live: loaded: MB/229.212.160
address 0xcc000000 out of loom!
loom: [0x36000000 : 0xb6000000)
Assertion '0' failed in ../noun/events.c:126

bail: oops
bailing out
urbit: ../noun/manage.c:650: u3m_bail: Assertion `0' failed.
Aborted (core dumped)
urb@nec:~$ ls
arvo  core-old  dump.txt    key.txt      nec        snap.snap   trace.txt
core  devops    ivory.pill  latest.snap  pill.pill  solid.pill  urbit
urb@nec:~$ lldb -c core
(lldb) target create --core "core"
Core file '/home/urb/core' (x86_64) was loaded.
(lldb) bt
* thread #1, name = 'urbit', stop reason = signal SIGABRT
  * frame #0: 0x00007f8e0e930077 libc.so.6`__GI_raise(sig=2) at raise.c:51
    frame #1: 0x00007f8e0e911535 libc.so.6`__GI_abort at abort.c:79
    frame #2: 0x00007f8e0e91140f libc.so.6`__assert_fail_base.cold.0 at assert.c:92
    frame #3: 0x00007f8e0e921142 libc.so.6`__GI___assert_fail(assertion=<unavailable>, file=<unavailable>, line=<unavailable>, function=<unavailable>) at assert.c:101
    frame #4: 0x000055acd7a089de urbit`u3m_bail + 414
    frame #5: 0x000055acd7a0883e urbit`c3_cooked + 14
    frame #6: 0x000055acd79ff41c urbit`u3e_fault + 380
    frame #7: 0x00007f8e0ed04e6c libsigsegv.so.2`___lldb_unnamed_symbol3$$libsigsegv.so.2 + 108
    frame #8: 0x00007f8e0f1ccdd0 libpthread.so.0`___lldb_unnamed_symbol1$$libpthread.so.0 + 1
    frame #9: 0x000055acd7a0121e urbit`u3h_free + 142
    frame #10: 0x000055acd7a05ab3 urbit`u3j_boot + 339
    frame #11: 0x000055acd7a0a516 urbit`u3m_boot + 102
    frame #12: 0x000055acd79dcd89 urbit`main + 1929
    frame #13: 0x00007f8e0e91309b libc.so.6`__libc_start_main(main=(urbit`main), argc=2, argv=0x00007ffd59ddc7d8, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007ffd59ddc7c8) at libc-start.c:308
    frame #14: 0x000055acd79dd3da urbit`_start + 42
(lldb)
```

**Comment by joemfb** [comment]
That's using the wrong urbit binary.

**Comment by jtobin** [comment]
On Thu, Jul 11, 2019 at 04:54:09PM -0700, Joe Bryan wrote:
> That's using the wrong urbit binary.

Make sure the nix environment is on your PATH to get at the rc binaries:

  source ~/.nix-profile/etc/profile.d/nix.sh

You might also want to use 'urbit-testnet', which has the |mass dumping
configured.

**Comment by joemfb** [comment]
The crash in the linked gist was caused by leaking the replacement event in the daemon. Fix in #1330.

**Comment by ixv** [comment]
> Make sure the nix environment is on your PATH to get at the rc binaries

Ah my bad, it started up just fine after this.

I just got another bail thats similar but different, on ~winsyx-wicsyl.

It's a bail foul this time, unsure if this is the same problem. Console output here, with core back trace at the bottom of the file: https://gist.github.com/ixv/ce42965e3f9c49026d536937c1fc196e

**Comment by joemfb** [comment]
This is not the same problem -- the backtrace is unrelated to `|mass` and `bail: foul` is a reference undercount. The ship was running without the changes in #1330, right (I want to make sure they're not implicated).

**Comment by ixv** [comment]
Right

**Comment by joemfb** [comment]
Note that this is the same memory leak, but a different crash -- it appears to be caused by the terminal spinner thread. The reason we abort when we discover a memory leak is that we've discovered it through garbage collection, which itself works by manipulating of the reference counts. When we've discovered a leak this way, we've also gone into an invalid memory state, so we can't keep working.

In other words, I think the difference in this crash is nonessential, and the underlying problem should still be fixed by #1330

**Comment by vvisigoth** [comment]
👑

**Comment by vvisigoth** [comment]
`~zod` bailed again on testnet. Here are the last lines in scrollback, nothing particularly interesting as far as I can see.

```
+ /~zod/home/27/web/publish/my-cool-blog/image-posting/~2019.7.13..13.48.00..638f/publish
-comment   
write failed
Assertion '0' failed in vere/term.c:35

bail: oops 
bailing out
urbit: noun/manage.c:663: u3m_bail: Assertion `0' failed.
                                                         Aborted (core dumped) 
```

I can't seem to find the core, so I'm not going to try to restart this. @ixv @joemfb any idea of where it would be?

**Comment by joemfb** [comment]
I looked briefly and also didn't see one. Maybe the terminal was missing `ulimit -c unlimited`.

FWIW, this crash is just an `assert(0)` after a terminal `write()` call returned an error. These errors were previously ignored; this change was introduced to suppress compiler warnings in #1251. I suspect this is too heavy-handed, there are undoubtedly terminal writes that can be allowed to fail, or certain error conditions that can be safely ignored.

**Comment by vvisigoth** [comment]
reset the ulimit and restarted. all is "well"

**Comment by jtobin** [comment]
> In other words, I think the difference in this crash is nonessential, and the underlying problem should still be fixed by #1330

Closing as I think this one has outlived its usefulness.
